pennylane/transforms/metric_tensor.py

```diff
@@ -292,12 +325,226 @@ def processing_fn(probs):
+    The block diagonal of the tensor is computed using the covariance matrix approach
+    implemented in :func:`.metric_tensor_cov_matrix`. In addition, we may extract the
+    factors for the second terms :math:`\langle \psi|\partial_j\psi\rangle`
+    of the block off-diagonal tensor from the tape results for the covariance matrix.
+    This means that in total only the tapes for the first terms of the block off-diagonal
+    are required in addition to ``metric_tensor_cov_matrix``.
+    """
+    # Prepare aux_wire
+    aux_wire = aux_wire or tape.num_wires
+    # Get tapes and processing function for the block diagonal metric tensor,
+    # as well as the generator observables and generator coefficients for each diff'ed operation
+    diag_tapes, diag_proc_fn, obs_list, coeffs = _metric_tensor_cov_matrix(tape, diag_approx=False)
+    graph = tape.graph
+    par_idx_to_trainable_idx = {idx: i for i, idx in enumerate(sorted(tape.trainable_params))}
+    layers = graph.iterate_parametrized_layers()
+    layers = [
+        LayerData(
+            layer.pre_ops,
+            layer.ops,
+            tuple(par_idx_to_trainable_idx[idx] for idx in layer[2]),
+            layer.post_ops,
+        )
+        for layer in layers
+    ]
+
+    first_term_tapes = []
+    ids = []
+    block_sizes = []
+    # Get all tapes for the first term of the metric tensor
+    for idx_i, layer_i in enumerate(layers):
+        block_sizes.append(len(layer_i.param_inds))
+        for layer_j in layers[idx_i + 1 :]:
+            _tapes, _ids = _get_first_term_tapes(tape, layer_i, layer_j, allow_nonunitary, aux_wire)
+            first_term_tapes.extend(_tapes)
+            ids.extend(_ids)
+
+    # Combine block diagonal and off-diagonal tapes
+    tapes = diag_tapes + first_term_tapes
+    # prepare block off-diagonal mask
+    mask = 1 - qml.math.block_diag([qml.math.ones((bsize, bsize)) for bsize in block_sizes])
+    # Required for slicing in processing_fn
+    num_diag_tapes = len(diag_tapes)
+
+    def processing_fn(results):
+        diag_res, off_diag_res = results[:num_diag_tapes], results[num_diag_tapes:]
+        # Get full block diagonal tensor
+        diag_mt = diag_proc_fn(diag_res)
+        # Initialize block off-diagonal tensor using the stored ids
+        first_term = qml.math.zeros_like(diag_mt)
+        for result, idx in zip(off_diag_res, ids):
+            # The metric tensor is symmetric
+            first_term = qml.math.scatter_element_add(first_term, idx, result[0])
+            first_term = qml.math.scatter_element_add(first_term, idx[::-1], result[0])
+
+        # Second terms of block off-diagonal metric tensor
+        expvals = []
+        for prob, obs in zip(diag_res, obs_list):
+            for o in obs:
+                l = qml.math.cast(o.eigvals, dtype=np.float64)
+                w = tape.wires.indices(o.wires)
+                p = qml.math.marginal_prob(prob, w)
+                expvals.append(qml.math.dot(l, p))

```

Comment:
so the block diagonal elements actually contain half the information needed to compute the off-diagonal elements?

Reply:
Depending on how one computes the "Half", yes :D It contains the information for the second term of the tensor, an outer product of a vector we already have. However, the first term dominates the cost on the off-diagonal!